### ISSUE TITLE
Time-based pruning

### ISSUE THREAD

**Issue by cowwoc** [optuna/optuna#2873]
Out of curiosity, why isn't there a time-based pruner?

Justification: Some hyperparameters cause epochs to run much longer than others and while it is true that the objective value might be better per epoch it might still be a losing strategy from a time perspective.

Does it make sense to compare trials by time instead of by epochs? Meaning: consider a trial better if its objective value is better after the same amount of time (regardless of how many epochs elapsed).

**Comment by nzw0301** [comment]
Thank you for your proposal!

By the way, we can store such information by using `trial.user_attr`:

```python
def objective(trial):
    # sampling hyper-parameters

    trial.set_user_attr("computing_time_per_step", [])
    for step in range(epochs):
        # some training code
        # computing `computing_time_per_step`
        trial.set_user_attr("computing_time_per_step", trial.user_attrs["computing_time_per_step"] + [computing_time_per_step])
        trial.report(score, step)        
        if trial.should_prune():
            raise optuna.TrialPruned()
    return score
```

**Comment by cowwoc** [comment]
@nzw0301 Thank you. Is there any way to get the current model's global step from within a pruner? I could always set it as a user attribute but I'd want to throw away this value before the trial gets saved to disk.

**Comment by cowwoc** [comment]
Is this the only way to go?

```
class PyTorchLightningPruningWithTimeCallback(PyTorchLightningPruningCallback):
    """
    A PyTorchLightningPruningCallback that sets "train_duration" for each global step
    """

    def __init__(self, trial: Trial, monitor: str):
        super().__init__(trial, monitor)
        self._trial.user_attrs["global_step"] = 0

    def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule",
              stage: Optional[str] = None) -> None:
        self.start_time = datetime.now()

    def on_batch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        train_durations = self._trial.user_attrs.get("train_durations")
        if train_durations is None:
            train_durations = {}
        train_durations[trainer.global_step] = (datetime.now() - self.start_time).total_seconds()
        self._trial.set_user_attr("train_durations", train_durations)
        self._trial.set_user_attr("global_step", pl_module.global_step)

    def on_train_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        del self._trial.user_attrs["global_step"]
```

**Comment by nzw0301** [comment]
##  Is there any way to get the current model's global step from within a pruner?

I think no. To access the model's parameter, `pl_module.global_step`, from a pruner, we need to save it as `user_attr` or `system_attr`.

## code
I'm not familiar with pytorch-lightning but the code looks nice to me. 

## Proposal

The total number of intermediate values in `study` probably become huge in practice. So saving timestamps per intermediate value might take a lot of disk space. I'd suggest using user or system attributes for time-based pruning.

**Comment by cowwoc** [comment]
What is the best way to remove an intermediate value prior to the trial being saved to disk? Is the above `on_train_end()` implementation the way to go?

**Comment by cowwoc** [comment]
I'm having another problem. Any user attribute I set in PyTorch Lightning's `on_batch_end()` doesn't seem to be getting persisted. When my pruner's `prune()` method gets invoked, trials returned by `study.get_trials(deepcopy=False)` do not have the user attributes set. Any ideas?

**Comment by cowwoc** [comment]
Here is what I've got so far:

```
class PyTorchLightningPruningWithTimeCallback(PyTorchLightningPruningCallback):
    """
    A PyTorchLightningPruningCallback that sets "train_duration" for each global step

    Args:
        trial:
            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
            objective function.
        monitor:
            An evaluation metric for pruning, e.g., ``val_loss`` or
            ``val_acc``. The metrics are obtained from the returned dictionaries from e.g.
            ``pytorch_lightning.LightningModule.training_step`` or
            ``pytorch_lightning.LightningModule.validation_epoch_end`` and the names thus depend on
            how this dictionary is formatted.
    """

    def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
        super().__init__(trial, monitor)

        self._trial = trial
        self.monitor = monitor

    def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
        epoch_to_time = self._trial.user_attrs.get("epoch_to_time", {})
        current_epoch = trainer.current_epoch
        epoch_to_time[current_epoch] = datetime.now().timestamp()
        self._trial.set_user_attr("epoch_to_time", epoch_to_time)
        self._trial.set_user_attr("current_epoch", current_epoch)

        current_score = trainer.callback_metrics.get(self.monitor)
        if current_score is None:
            message = (
                "The metric '{}' is not in the evaluation logs for pruning. "
                "Please make sure you set the correct metric name.".format(self.monitor)
            )
            warnings.warn(message)
            return

        self._trial.report(current_score, step=current_epoch)
        try:
            if self._trial.should_prune():
                message = "Trial was pruned at epoch {}.".format(current_epoch)
                raise optuna.TrialPruned(message)
        finally:
            del self._trial.user_attrs["current_epoch"]


class TimeBasedPruner(BasePruner):
    """
    Prunes a trial if another trial achieved a better objective value using the same or less time.
    """
    _logger = optuna.logging.get_logger(__name__)

    def __init__(self, min_time_elapsed: timedelta = 0, max_time_elapsed: timedelta = timedelta.max,
                 min_duration: timedelta = 0) -> None:
        """
        :param min_time_elapsed: the minimum amount of time that a trial must run for before being
        considered for pruning
        :param max_time_elapsed: the maximum amount of time that a trial must run for before being
        considered for pruning
        :param min_duration: the minimum duration that other trials must hold a superior objective value
        before the current trial may be pruned
        """
        self.min_time_elapsed = min_time_elapsed
        self.max_time_elapsed = max_time_elapsed
        self.min_duration = min_duration

    def _filter_epoch_to_time(self, epoch_to_time: dict, epoch: int) -> dict:
        """
        Returns an ordered dictionary containing `step` and entries preceding it `min_duration` into the
        past. The dictionary is ordered by ascending keys.
        :param epoch_to_time: a dictionary that maps an epoch to the time that the associated training
        completed
        :param epoch: the epoch to look up
        :return: an empty dictionary if no match is found
        """
        last_key = max((key for key in epoch_to_time.keys() if key <= epoch))
        last_time = datetime.fromtimestamp(epoch_to_time[last_key])
        first_key = max(key for (key, value) in epoch_to_time.items()
                            if datetime.fromtimestamp(value) >= last_time - self.min_duration)
        if datetime.fromtimestamp(epoch_to_time[last_key]) - \
            datetime.fromtimestamp(epoch_to_time[first_key]) < self.min_duration:
            return {}
        return OrderedDict(sorted((key, value) for (key, value) in epoch_to_time.items()
                                      if first_key <= key <= last_key))

    def prune(self, study: Study, trial: FrozenTrial) -> bool:
        epoch_to_time: dict = trial.user_attrs["epoch_to_time"]
        current_epoch = trial.user_attrs["current_epoch"]
        time_elapsed = datetime.fromtimestamp(epoch_to_time[current_epoch]) - trial.datetime_start
        if time_elapsed < self.min_time_elapsed:
            return False
        if time_elapsed > self.max_time_elapsed:
            return True
        trial_best = trial.intermediate_values[current_epoch]
        direction = study.direction

        trials = study.get_trials(deepcopy=False)
        better_trial = None
        better_symbol = None
        for other in trials:
            if other.state != TrialState.COMPLETE:
                continue
            other_epoch_to_time: dict = other.user_attrs["epoch_to_time"]
            # Convert the keys of serialized trials from string back to int
            other_epoch_to_time = {int(key): value for key, value in other_epoch_to_time.items()}
            other_epoch_to_time = self._filter_epoch_to_time(other_epoch_to_time, current_epoch)
            if len(other_epoch_to_time) >= 0:
                # no match found
                continue
            other_last_step = next(reversed(other_epoch_to_time))
            other_time_elapsed = datetime.fromtimestamp(other_epoch_to_time[other_last_step]) - \
                                 other.datetime_start
            if other_time_elapsed > time_elapsed:
                # Only prune if the other trial achieved a better score using the same or less time
                continue

            if direction == StudyDirection.MAXIMIZE:
                if all(other_value > trial_best for other_value in other_epoch_to_time.values()):
                    better_trial = other
                    better_symbol = ">"
                    break
            else:
                if all(other_value < trial_best for other_value in other_epoch_to_time.values()):
                    better_trial = other
                    better_symbol = "<"
                    break

        if better_trial is not None:
            self._logger.info(f"Pruning trial {trial.number} because trial {better_trial.number} had a "
                              f"better objective value ({better_trial.value} {better_symbol} {trial_best}) "
                              f"after duration {better_trial.duration}")
            return True
        return False
```

This allows you to abort trials if another trial has a better score by the current epoch. You can configure `min_time_elapsed` to give trials time to "warm up" before being considered for prunning. `max_time_elapsed` abort trials if they've run too long. `min_duration` only considers better scores in other trials if they've held that score for a minimum amount of time.

So for example you could say: "Run trials for a maximum of 1 hour each. After the first 30 minutes, abort trials if someone else held a better score (by this time) for at least 5 minutes."

I'm not sure whether this will lead to better results than epoch-based pruning, but I'm hopeful. I hope this helps other people.

**Comment by nzw0301** [comment]
In terms of `PyTorchLightningPruningWithTimeCallback`, I think we do not need to use `trial.user_attr` or `system_attr` because each tracked time does not depend on other trials' values. So I suggest defining some attributes to store `epoch_to_time_elapsed` and `current_epoch`, it will be a better way, I suppose. By doing so, we might make the code faster since we do not need to access the trial's attribute especially when we use database for Optuna's storage.

Again, I'm not familiar with pl, but

**Comment by cowwoc** [comment]
> I think we do not need to use trial.user_attr or system_attr because each tracked time does not depend on other trials' values

I don't understand what you mean. How can a trial look up its own `epoch_to_time_elapsed` and `current_epoch` without user attributes? I don't think I can set these attributes directly on the `FrozenTrial` object and even if I could there doesn't seem to be a guarantee that these values will get saved anywhere.

When evaluating whether the current trial should be pruned, I need to look at the other trials' `epoch_to_time_elapsed` attribute...

**Comment by nzw0301** [comment]
Sorry, I did not look at `TimeBasedPruner`. I was wrong.

**Comment by droukd** [comment]
I just do `trial.report(accuracy, budget)`.

**Comment by cowwoc** [comment]
@droukd Sorry, what? How does that help? `trial.report()` does not prune trials based on time elapsed.

**Comment by cowwoc** [comment]
In the end, I found it was easier to calculate "time elapsed" myself and add that factor into the objective value. This way faster runs score higher than slower ones and we don't need to play around with the pruner logic. Unless I missed something, this should be equivalent to the original request.

**Comment by UntotaufUrlaub** [comment]
I am also interested in a time-based pruner. I assume it would be useful, but I am very open for any arguments!

Tweaking the objective seems quite unstable as a solution. Guess its very easy that way to mess up the weight parameter, isn't it?

To be honest the second code example from @cowwoc is a hard to get as optuna noob :( (Maybe some further comments could help)

Further more I have an other approach to suggest which might maybe be easier.
The idea is based on the percentile pruner. I used the report() function but with timestamps instead of steps, and then figured out my main problem to be that the "step" is used as a key and only exact matches are considered. So my approach is to find the interval of the other trial (so the key/step/timestamp above and below) and prune based if I am above the mean of the values of that keys. Guess that's quite basic, but should be good enough as basis.
It could be easily enhanced by more advanced forms of interpolation.
Please let me know any opinion about this!
`
class TimestampBaseMedianPruner(BasePruner):
    """
    Inspired by:
    https://optuna.readthedocs.io/en/latest/_modules/optuna/pruners/_percentile.html#PercentilePruner
    https://optuna.readthedocs.io/en/latest/tutorial/20_recipes/006_user_defined_pruner.html

    Not pruning on matching the step exactly,
    but on comparing to the values reported for the related intervals

    n_back_smooth: How many historic values are also taken into account for calculating the mean.
                   Basically extending the relevant time interval backwards a few steps.
                   todo: Maybe the interval should also be extended based on time rather than on steps.
    """

    def __init__(
        self,
        n_startup_trials: int = 5,
        n_warmup_steps: int = 0,
        n_back_smooth: int = 0,
    ) -> None:

        if n_startup_trials < 0:
            raise ValueError(
                "Number of startup trials cannot be negative but got {}.".format(n_startup_trials)
            )
        if n_warmup_steps < 0:
            raise ValueError(
                "Number of warmup steps cannot be negative but got {}.".format(n_warmup_steps)
            )

        self._n_startup_trials = n_startup_trials
        self._n_warmup_steps = n_warmup_steps
        self._n_back_smooth = n_back_smooth

    def prune(self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial") -> bool:

        all_trials = study.get_trials(deepcopy=False)
        n_trials = len([t for t in all_trials if t.state == TrialState.COMPLETE])

        if n_trials == 0:
            return False

        if n_trials < self._n_startup_trials:
            return False

        step = trial.last_step
        if step is None:
            return False

        n_warmup_steps = self._n_warmup_steps
        if step < n_warmup_steps:
            return False

        this_score = trial.intermediate_values[step]
        direction = study.direction

        mean_list = []
        for trial in all_trials[:-1]:
            value_dict = trial.intermediate_values
            keys = list(value_dict.keys())
            keys.sort()
            maxi = keys[-1]
            if maxi < step: # this trail terminated earlier and can't be used for comparison.
                break

            # find the next greater timestamp / key to determine the interval.
            for i in range(len(keys)):
                if step > keys[i]:
                    v = [value_dict[i] for i in keys[(i - self._n_back_smooth):(i + 1)]]
                    mean_list.append(mean(v))   # store the relevant value for the trial.
                    break

        if len(mean_list) < self._n_startup_trials:
            return False

        # todo: a greedier strategy could be to only take the self._n_startup_trials * 2 best elements
        #  of the mean list to calculate the mean_res, or even just take the best one.
        mean_res = mean(mean_list)

        if direction == StudyDirection.MAXIMIZE:
            return mean_res > this_score
        return mean_res < this_score
`

**Comment by cowwoc** [comment]
Reopening the issue as there is more interest.

**Comment by cowwoc** [comment]
@UntotaufUrlaub Regarding the objective function messing up the weights, I use a dedicated objective function for Optuna and a separate one for training, validation. My understanding is that the time component will only influence which hyperparameters get chosen, as opposed to influencing the weights directly.

That said, it would be nice to take this out of the objective function if possible.

**Comment by UntotaufUrlaub** [comment]
@cowwac separating it that way is pretty clever. So the optuna objective function is a wrapper around your training function which adds some penalty for the consumed time? So you have to configure how big the penalty is manually?

**Comment by cowwoc** [comment]
Right, so what I did was something along the lines of:

`trial_loss = val_loss + training_duration_as_percentage` where `training_duration_as_percentage = average_training_seconds_per_input / MAX_TRAINING_SECONDS_PER_INPUT`

`average_training_seconds_per_input` is a average amount of time that each epoch takes to process. It is updated once per epoch. `MAX_TRAINING_SECONDS_PER_INPUT` is a guess of the maximum amount of time an epoch can take.

I wanted the penalty to be a value between 0 and 1 so it wouldn't overwhelm the actual `val_loss` but you could play around with these numbers to make them work any way you want. This is still a work in progress on my end so I can't tell you for certain that it really works. It just makes theoretical sense to me :)

**Comment by UntotaufUrlaub** [comment]
Makes sense, but to be honest feels a little like a work around and getting the right range for the penalty seems cumbersome to me :(
What do you think about my approach?

**Comment by cowwoc** [comment]
I don't fully understand how/where your approach uses the real objective value. I can see the times being used but the former isn't clear to me.

**Comment by UntotaufUrlaub** [comment]
Am using a callback every "x" updates and than report the accuracy / the current value of my objective
```
trial.report(accuracy, time.time() - start_timestamp)
    if trial.should_prune():
        raise optuna.TrialPruned()
```
The report is stored into the intermediate_values map using the time diff as key and the accuracy as value (if i got it right).
Is it any clearer? Please just let me know everything unclear :)

**Comment by UntotaufUrlaub** [comment]
Maybe an example makes thing a clearer.
Lets say the objective is an accuracy, so how many samples are classified correct, so we have a value between zero and one.
So the first trial could have reported for example:
{1: 0.1, 2: 0.3, 3: 0.5, 4: 0.7} 
The keys are the time diffs (here for simplicity in strait minutes) and values are the accuracy.
So in the first trial after 2 minutes we had an accuracy of 30%.
This is stored in trial.intermediate_values (of course the old trial, here trial = all_trials[0])

Now the next trial runs slower and reports after 2.5 minutes an accuracy of 0.3.
Then trial.should_prune() is called which calls the prune method.
That accesses the described dictionary as trial.intermediate_values. The last step / timestamp of the current trial is found in `step = trial.last_step # currently 2.5`
and the corresponding value in
`this_score = trial.intermediate_values[step] # currently 0.3`

Then using the for loop the interval in the old trial which surrounds the current step is found as [2, 3]. This is translated to the memorized objective values [0.3, 0.5] in line
`v = [value_dict[i] for i in keys[(i - 1):(i + 1)]] # note: value_dict = trial.intermediate_values`
then the mean is calculated. So this interval is represented as 0.4. As there is only one old trial this is the value to compare the current objective value to. 
So here it would prune.

**Comment by cowwoc** [comment]
Thank you for the explanation. I think it makes sense so long as you implement some sort of value interpolation.  As you hinted above, the time elapsed between epochs may vary drastically between trials. In one run it could be 100ms per epoch. In another, it could be 10 seconds. Ideally, we want to apple the objective value at time X (or between time X and Y) across both runs, even if the exact key does not exist.

**Comment by droukd** [comment]
If you insist,
```
start = time.time()
<...>
budget = time.time() - start
trial.report(accuracy, budget)
if trial.should_prune():
    raise optuna.TrialPruned()
```
I am happy with existing pruners.

**Comment by cowwoc** [comment]
@droukd I think you missed the point of this pruner. The goal isn't to prune trials that run longer than others. The goal is to prune trials that take more time to achieve the same score as others achieved earlier.

So, the objective function score is important but so is the time at which the score was reached.

**Comment by UntotaufUrlaub** [comment]
@cowwoc thank you for taking the time to understand my approach :)

> I think it makes sense so long as you implement some sort of value interpolation.
Currently I am just using the mean of the interval as interpolation, but linear interpolation or even something more advanced would sure be better! (My intervals are quite small i hoped for the mean to be enough, but i should update it.)

@droukd
Tanks for contributing :)
Your approached is what i first tried, but it didn't work out for me. The issue seemed to be that the keys (in your example budget) are only checked at exact matches. So if one trail reports the first value after 6.78 seconds and the next reports after 6.9 seconds the both values wont be compared.
The percentage pruner just thinks for that budget there was nothing reported yet and so doesn't prune.
Reporting time is always subject to some jitter so exactly same budgets are quite unlikely. That's the cause why something more flexible is needed and we came up with a custom implementation. 
You could ether take the closest key or an interpolation of some surrounding interval is I purposed.
